### DISCUSSION TITLE
3.1.0

### DISCUSSION THREAD

**Discussion by marcus-j-davies** [zwave-js/ZWaveJS.NET#23] [General]
Hi @georgeinva2004

Any ill health to report with 3.1.0? - I'm tempted to release the new version this weekend (maybe even this evening), if 3.1.0 is working ok. it didn't need to address many issues, more of a tightening up exercise.

**Comment by georgeinva2004** [comment]
Sorry for the late response.  I thought I had hit send on my email.  No issues.  
Thanks for the great work!!!!

**Comment by georgeinva2004** [comment]
Just installed the prod version of 3.1.0.  Did anything get changed on ValueNotification events between the alpha and the prod version?  They seem to have stopped working.  The driver appears to be getting  the events from the device, but I'm not seeing the ValueNotification event firing.  I'm forcing the download of the new PSI.    Other events appear to be working.


----------------------------------------Driver log-------------------------------------------------------------
2022-06-14T16:57:04.116Z SERIAL « 0x010d00040024055b03488001cc008a                                    (15 bytes)
2022-06-14T16:57:04.117Z SERIAL » [ACK]                                                                   (0x06)
2022-06-14T16:57:04.119Z DRIVER « [Node 036] [REQ] [ApplicationCommand]
                                  └─[CentralSceneCCNotification]
                                      sequence number: 72
                                      key attribute:   KeyPressed
                                      scene number:    1
2022-06-14T16:57:04.121Z CNTRLR   [Node 036] [!] [Central Scene] scene[001]: 0
2022-06-14T16:57:04.121Z CNTRLR « [Node 036] received CentralScene notification {
                                      "nodeId": 36,
                                      "ccId": "Central Scene",
                                      "ccCommand": "0x03",
                                      "payload": "0x488001"

**Reply by marcus-j-davies** [reply]
Hi @georgeinva2004,

Mmm..
Nothing should be different here.
Although, the latest PSI  uses ZWave JS 9.4.0 - But that shouldn't present any issues.

https://github.com/zwave-js/node-zwave-js/releases/tag/v9.4.0

I assume you're  subscribing to `ValueNotification` on the `ZWaveNode` class?

is the connection still live after this attempt?

**Reply by marcus-j-davies** [reply]
If you debug your app - open up the `output` or `debug` window I'm sure I put out the incoming payload. 

https://github.com/zwave-js/ZWaveJS.NET/blob/1e0b852dae990600d4e03a2a50d67fc98918bb8d/Visual%20Studio%20Projects/ZWaveJS.NET/ZWaveJS.NET/Driver.cs#L364

**Reply by georgeinva2004** [reply]
Very strange.  I'm not seeing any payload.  However, I excluded and included the device and the event is now getting fired.    Never had to do that before :)

**Reply by marcus-j-davies** [reply]
Very strange.

Although, an area of interest has my attention.

I immediately  start the `zwave-js-server`, once the driver becomes ready (this lives inside the PSI Image) 
https://github.com/zwave-js/ZWaveJS.NET/blob/1e0b852dae990600d4e03a2a50d67fc98918bb8d/PSI/server.js#L31



I wonder if I should allow a little time, before opening up the driver to the server.
Our WS client will retry every 100ms - and of course this succeeds once the driver is ready - It maybe things are not 'quite ready' for primetime - and causing some issues later.

so we do this.

```javascript
driver.on("driver ready", () => {
   setTimeout(()=>{
       server.start();
       ServerStarted = true;
   }, 1000);
});
```

Let me know if you see a pattern, and I may add some delay between the driver being ready and the server accepting connections.

the driver being ready, is different from nodes being ready - so if you continue to see oddities like this, a delay may be needed during start up (but after Driver ready - allowing nodes to follow suite) - keep me updated. 👍

**Reply by georgeinva2004** [reply]
So far, things are okay.  I suppose the delay wouldn't hurt.  I wait until I get the DriverReady event before I actually do anything with the driver.  The first thing I do on startup (after the driver is ready) is enumerate nodes and subscribe to node-related events. All that seems to be working now.  I am noticing something weird with a device that implements the Indicator command class.  I can control the indicator state through my app.  However, when I press the physical button that changes the Indicator state (e.g. turns the indicator off), I don't get a ValueNotification or ValueUpdated notification associated with the Indicator value ( I do get a ValueNotification event for the Central Scene Value).  Further, when I read the value of the of the Indicator (on/off Boolean state) it always returns true even if the indicator is visually off.  Not sure if this is a driver issue or a device issue or a "me not understanding Z-Wave" issue.  

This is the value id associated with the Indicator:
{
		"commandClass": 135,
		"endpoint": 0,
		"property": 80,
		"propertyKey": 2,
		"commandClassName": "Indicator",
		"propertyName": "Node Identify",
		"propertyKeyName": "Binary"
	}

**Reply by marcus-j-davies** [reply]
Yup, the Indicator class doesn't actually have a 'hey, I was just updated' event.

I don't know of any device that actually emits the value updated event for indicators.

You should be able to call the method of PollValue on that ValueID - this forces the device to be queued for it - therefore updating the drivers local cache.

I have Security keypad, that also doesn't emit a value when it's indicator is updated (I checked this with a zwave sniffer). So it seems quite normal.

@AlCalzone,
Indicator class doesn't have an event mechanism right? But local values should reflect on the successful set? > V1 according to the ValueID above, where it's a Boolean indicator.

@georgeinva2004 
I think the delay (if we need it) is more the contract between zwave-js-server and the driver, rather then anything the user needs to worry about.

**Reply by georgeinva2004** [reply]
PollValue did the trick.   Now I just need to reliably correlate the Indicator propertyName to the Central Scene propertyKeyName and I'll be in business.  Thanks for the tip!

**Reply by marcus-j-davies** [reply]
Nice!

The difference between Get and Poll value, is that Get fetches the value from the drivers cache (which is usually kept up to date, as and when things change)

The cache also survives restarts - it writes it to disk.

Poll on the other hand, forces a refetch of that value from the device.

Poll can be used, where you have had some downtime maybe, where the cached values may not be current.

**Reply by AlCalzone** [reply]
> Indicator class doesn't have an event mechanism right? But local values should reflect on the successful set?

Correct. There is no value notification for the Indicator CC, only value added/updated events.

And a Central Scene notification does not influence Indicator values, because there is nothing in the specs that says it should. If the device changes its indicator values in tandem and does not send a corresponding report, "manually" refreshing them is necessary.

**Reply by georgeinva2004** [reply]
You ever have one of those days where you are overthinking a problem?  Well, that was me yesterday.  After reading @AlCalzone 's response it occurred to me that using the Indicator command class to determine Scene state was a bad assumption, and would have required more work than was required.  Following the KISS principle, I'm simply going to force the Indicator (if the command class exists) to sync with what my controlling app has defined as the scene state.  Thanks for the help folks!!!!

**Reply by marcus-j-davies** [reply]
Everyday!

**Comment by georgeinva2004** [comment]
No issues. Very stable.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Marcus Davies ***@***.***>
Sent: Friday, June 10, 2022 9:41:45 AM
To: zwave-js/ZWaveJS.NET ***@***.***>
Cc: John M. George ***@***.***>; Mention ***@***.***>
Subject: [zwave-js/ZWaveJS.NET] 3.1.0 (Discussion #23)


@georgeinva2004<https://github.com/georgeinva2004>

Any ill health to report with 3.1.0? - I'm tempted to release the new version this weekend (maybe even this evening), if 3.1.0 is working ok. it didn't need to address many issues, more of a tightening up exercise.
